ivm/src/ext.rs

```diff
@@ -5,13 +5,225 @@
 
 use core::{
   fmt::{self, Debug},
-  mem::transmute,
+  marker::PhantomData,
   ops::{Add, Div, Mul, Sub},
 };
-use std::io::{self, Read, Write};
+use std::{
+  collections::HashMap,
+  io::{self, Read, Write},
+};
 
 use crate::port::Tag;
 
+macro_rules! define_ext_fns {
+  ($map:ident,$self:ident, $($name:expr => |$a:ident, $b: ident| $body:expr ),*) => {
+    $($map.insert($name.to_string(), $self.register_ext_fn(move |[$a, $b]| [$body]).unwrap());)*
+  };
+}
+#[derive(Default)]
+pub struct Extrinsics<'ivm> {
+  ext_fns: Vec<Box<dyn Fn(ExtVal<'ivm>, ExtVal<'ivm>) -> ExtVal<'ivm> + Sync + 'ivm>>,
+  // amount of registered light exttys
+  light_ext_ty: u16,
+
+  n32_ext_ty: Option<ExtTy<'ivm>>,
+  io_ext_ty: Option<ExtTy<'ivm>>,
+
+  phantom: PhantomData<fn(&'ivm ()) -> &'ivm ()>,
+}
+
+impl<'ivm> Extrinsics<'ivm> {
+  pub const MAX_EXT_FN_KIND_COUNT: usize = 0x8FFF;
+  pub const MAX_LIGHT_EXT_TY_COUNT: usize = 0x8FFF;
+  pub const MAX_RC_EX_TTY_COUNT: usize = 0x8FFF;
+
+  pub fn register_ext_fn(
+    &mut self,
+    f: impl Fn([ExtVal<'ivm>; 2]) -> [ExtVal<'ivm>; 1] + Sync + 'ivm,
+  ) -> Option<ExtFn<'ivm>> {
```

Comment:
I don't think there's much useful to do if this returns `None`, so all callers are likely to unwrap this. Can this panic, instead of returning an option?

Reply:
same for `register_light_ext_ty` below